Sources/Plasma/PubUtilLib/plPhysX/plPXPhysical.cpp

```diff
@@ -791,601 +463,211 @@ void plPXPhysical::ClearLinearVelocity()
-                                                tris.AcquireArray(),
-                                                mat,
-                                                l2w,
-                                                blended,
-                                                &idx,
-                                                myDraw);
-            }
-        }
+        physx::PxBoxGeometry geometry;
+        shape->getBoxGeometry(geometry);
+        return IGenerateProxy(addTo, idx, shape, geometry, l2w, mat, blended);
     }
 
-    NxConvexShape* convexShape = shape->isConvexMesh();
-    if (convexShape)
+    case physx::PxGeometryType::eCONVEXMESH:
     {
-        NxConvexMeshDesc desc;
-        convexShape->getConvexMesh().saveToDesc(desc);
-
-        hsTArray<hsPoint3>  pos;
-        hsTArray<uint16_t>    tris;
-
-        pos.SetCount(desc.numVertices);
-        tris.SetCount(desc.numTriangles * 3);
-
-        for (int i = 0; i < desc.numVertices; i++ )
-            pos[i] = GetConvexVert(desc, i);
-
-        for (int i = 0; i < desc.numTriangles; i++)
-            GetConvexTri(desc, i, &tris[i*3]);
-
-        myDraw = plDrawableGenerator::GenerateDrawable(pos.GetCount(), 
-            pos.AcquireArray(),
-            nil,    // normals - def to avg (smooth) norm
-            nil,    // uvws
-            0,      // uvws per vertex
-            nil,    // colors - def to white
-            true,   // do a quick fake shade
-            nil,    // optional color modulation
-            tris.GetCount(),
-            tris.AcquireArray(),
-            mat,
-            l2w,
-            blended,
-            &idx,
-            myDraw);
+        physx::PxConvexMeshGeometry geometry;
+        shape->getConvexMeshGeometry(geometry);
+        return IGenerateProxy(addTo, idx, shape, geometry, l2w, mat, blended);
     }
 
-    NxSphereShape* sphere = shape->isSphere();
-    if (sphere)
+    case physx::PxGeometryType::eSPHERE:
     {
-        float radius = sphere->getRadius();
-        hsPoint3 offset = plPXConvert::Point(sphere->getLocalPosition());
-        myDraw = plDrawableGenerator::GenerateSphericalDrawable(offset, radius,
-            mat, l2w, blended,
-            nil, &idx, myDraw);
+        physx::PxSphereGeometry geometry;
+        shape->getSphereGeometry(geometry);
+        return IGenerateProxy(addTo, idx, shape, geometry, l2w, mat, blended);
     }
 
-    NxBoxShape* box = shape->isBox();
-    if (box)
+    case physx::PxGeometryType::eTRIANGLEMESH:
     {
-        hsPoint3 dim = plPXConvert::Point(box->getDimensions());
-        myDraw = plDrawableGenerator::GenerateBoxDrawable(dim.fX*2.f, dim.fY*2.f, dim.fZ*2.f,
-            mat,l2w,blended,
-            nil,&idx,myDraw);
+        physx::PxTriangleMeshGeometry geometry;
+        shape->getTriangleMeshGeometry(geometry);
+        return IGenerateProxy(addTo, idx, shape, geometry, l2w, mat, blended);
+    }
+
+    DEFAULT_FATAL(shape->getGeometryType())
```

Comment:
I get a warning "control reaches end of non-void function" for this function, but in theory the `DEFAULT_FATAL` should make that not true?